### ISSUE TITLE
Plans for language evolution ?

### ISSUE THREAD

**Issue by lutter** [mruby/mruby#3481]
I just ran across mruby and am really excited about it. Currently, mruby implements the Ruby 1.9 language. Are there plans for adding more language features so that it will one day implement the same language as MRI ? If there are such plans, are there any timeframes for it ?

**Comment by felixjones** [comment]
About Ruby 2.0 support https://github.com/mruby/mruby/issues/2062
> Gradually. [...] But unless contributed, the priority is low.

Keyword arguments (Ruby 2.0 feature) is something that has been discussed a few times. Recently, it was discussed in relation to it being part of the 1.3.0 release https://github.com/mruby/mruby/issues/3140#issuecomment-280707108

The current situation for mruby is to be compatible with [ISO standard Ruby](https://www.iso.org/standard/59579.html), from my understanding this means one of the current goals is to remain fully compatible with ISO/IEC 30170:2012, which means no language additions that would break compliance.

**Comment by r-obert** [comment]
i don't think new language features are that important.
for language features ruby1.8 was all i ever needed, only the implementation was sometimes an issue.

**Comment by lutter** [comment]
It's mostly a question of how comfortable it is for users to go back and forth between mruby and MRI - the situation for which I am thinking about mruby is that we have a lot of people familiar with MRI. Anything that lowers the friction in going between MRI and mruby will help in gaining acceptance for that use.

**Comment by matz** [comment]
@lutter The major difference between CRuby2.4 and mruby is refinements and keyword arguments. I have no plan to implement refinements. Keyword arguments are postponed to mruby 1.4.

**Comment by lutter** [comment]
@matz thanks for the quick reply. Is it fair to say that differences between mruby and CRuby (especially when it comes to what libraries/facilities are available) are mostly due to nobody having done the work yet rather than that there would be opposition in principle ? I am mostly trying to gauge where mruby is likely to go in the future, and not for any sort of commitment/plans to do work ...

**Comment by matz** [comment]
@lutter Yes, basically. Besides that, there are a few features in Ruby2.x that would consume memory (either program size or runtime memory) too much. We try to make mruby compact.

**Comment by lutter** [comment]
@matz THanks ! Makes perfect sense.

**Comment by shevegen** [comment]
I think it may be useful to keep this information also within mruby distribution itself, such as in the doc/ subdirectory.

E. g. right now there are some guides:

https://github.com/mruby/mruby/tree/master/doc/guides

Or actually perhaps in this file here:

https://github.com/mruby/mruby/blob/master/doc/limitations.md

This file already mentions a lot of differences as-is.

What could be added to this file may be, from matz:

I try to re-word the above:

"Because mruby attempts to remain compact, small and fast, there are a few features available in Ruby2.x that will not be available in mruby, in particular when they would consume too much memory - either from the program size or runtime memory".

(Perhaps also add a specific example.)

The other statement by matz here was:

"refinements and keyword arguments. I have no plan to implement refinements. Keyword arguments are postponed to mruby 1.4."

I believe that this could also be mentioned. Let me try to reword this into a possible standalone-addition, it is just an example. :)

"Refinements are not slated for addition to mruby."

and:

"Keyword arguments may be added to mruby at 1.4 or later."

(Actually I also feel that refinements are experimental in MRI too, or at the least it feels that way. I remember that presentation called "nobody uses refinements" or something like that, which also reminds me about the older presentation about "nobody knows nobu" :D )

Anyway, the above is of course only a suggestion, largely because lutter asked the above, and the issue here can probably be closed in my opinion, preserving the information. Thanks!

**Comment by beoran** [comment]
I agree that keyword arguments would be nice and that refinements are not needed for mruby. Actually, I would say that the ruby language as such is feature complete and needs no new grammar.  For me ISO compliance is a must.

**Comment by letarg0** [comment]
turn on/off:

- file acess mruby-io
- network
- GC

**Comment by matz** [comment]
Closing — "language evolution plans" meta-question from 2017. mruby has since implemented major Ruby 3.x syntax (keyword args, pattern matching, endless methods, frozen string literals, etc.). doc/mruby*.md covers per-release additions.
